### PR TITLE
Adding Randomized Singular Value Decomposition

### PR DESCRIPTION
**Pull request by Rbravo555**
This PR adds a fast method to calculate the Singular Value Decomposition of large matrices with relatively low rank. As described in Halko et al 2009 https://arxiv.org/abs/0909.4061

### CODE REVIEW

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
+import numpy as np
+
+class RandomizedSingularValueDecomposition():
+
+    def __init__(self, COMPUTE_U=True, COMPUTE_V=True, RELATIVE_SVD=True, USE_RANDOMIZATION=True):
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:19]: please add some comments to explain how this should be used and waht the parameters mean

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        self.COMPUTE_U = COMPUTE_U
+        self.COMPUTE_V = COMPUTE_V
+        self.RELATIVE_SVD = RELATIVE_SVD
+        self.USE_RANDOMIZATION = USE_RANDOMIZATION
+
+    def Calculate(self, A, e0 = 0):
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py]: doc

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+                    e=np.sqrt(e0**2 - eORTH**2)
+                else:
+                    e=e0
+                self.RELATIVE_SVD = 0
+                self.SVD_MaxSize = max(np.shape(A))
+                U,S,V,eSVDb = self._SingularValueDecompostionTruncated(B, e )
+                U = Q @ U
+                eSVD = np.sqrt(eORTH**2 + eSVDb**2)
+
+        return U,S,V,eSVD
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:61]: doc of the meaning of the ouputs

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+    def _RandomizedOrthogonalization(self, C , mu=0, R=0):
+
+        M,N=np.shape(C) # C has dimensions M by N
+        c = nC = np.linalg.norm(C, 'fro')  # Norm of the initial residual
+        if mu==0:
+            mu = max(M,N)*np.finfo(float).eps*nC/2  # Machine presicion parameter
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py]: typo

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+            if Rest >= N:
+                Q='FULL'; B=np.array([]); break
+
+            i+=1
+            R_old = R; nC_old = nC; R = Rest
+
+        return Q, B, nC, c
+
+    def _SingularValueDecompostionTruncated(self, B, epsilon = 0):
```
Comment (RiccardoRossi) [inline applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:163]: doc

File: applications/RomApplication/tests/test_randomized_singular_value_decomposition.py [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+def synthetic_matrix(degree, rows = 100,repetitions=20):
+    TestMatrix = np.zeros((rows,degree))
+    x = np.linspace(0,1,rows)
+    for i in range(degree):
+        TestMatrix[:,i] = np.power(x,i)
+    return np.matlib.repmat(TestMatrix, 1, repetitions)
+
+class TestRandomizedSVD(KratosUnittest.TestCase):
+
+    def test_radomized_svd(self):
```
Comment (RiccardoRossi) [inline applications/RomApplication/tests/test_randomized_singular_value_decomposition.py:23]: will need to skip if numpy is not found

**Review by RiccardoRossi — COMMENTED** [review]
overall i am positive, with two minor comments:

1 - we should skip test in the case numpy is not found
2 - please add a minimal doc about the expected inputs and outputs of the functions

File: applications/RomApplication/tests/test_randomized_singular_value_decomposition.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
+#import python class test
+import KratosMultiphysics.KratosUnittest as KratosUnittest
+
+#import python packages
+from KratosMultiphysics.RomApplication.randomized_singular_value_decomposition import RandomizedSingularValueDecomposition
+import numpy as np
+import numpy.matlib
+
+def synthetic_matrix(degree, rows = 100,repetitions=20):
```
Comment (RiccardoRossi) [inline applications/RomApplication/tests/test_randomized_singular_value_decomposition.py:13]: just ensure that the test runs very fast, even on a very slow machine

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py (hunk 1 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:19]: done

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py (hunk 2 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py]: added

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py (hunk 3 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:61]: done

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py (hunk 4 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py]: fixed

File: applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py (hunk 5 above)
Comment (Rbravo555) [reply applications/RomApplication/python_scripts/randomized_singular_value_decomposition.py:163]: added

File: applications/RomApplication/tests/test_randomized_singular_value_decomposition.py (hunk 6 above)
Comment (Rbravo555) [reply applications/RomApplication/tests/test_randomized_singular_value_decomposition.py:23]: skipped if no numpy

File: applications/RomApplication/tests/test_randomized_singular_value_decomposition.py (hunk 7 above)
Comment (Rbravo555) [reply applications/RomApplication/tests/test_randomized_singular_value_decomposition.py:13]: Ran it on my potato computer, and the test for the randomized svd takes about 0.05 seconds. 

(the already-in-place tests which launch a ROM simulation and compare its solution with the expected one, do take about 4 seconds in total on this very limited computer)

**Review by RiccardoRossi — APPROVED** [review]

**Pull request merged**